system/Autoloader/Autoloader.php

```diff
@@ -312,6 +312,7 @@ protected function discoverComposerNamespaces()
         $paths    = $composer->getPrefixesPsr4();
         $classes  = $composer->getClassMap();
 
+        $composer->unregister();
```

Comment:
This is in the `discoverComposerNamespaces()` method.
So when `$modules->discoverInComposer` is false, it does not run.
When `$modules->discoverInComposer` is false, Composer autoloader is still enabled.
Is it intentional?

Reply:
Composer's autoloader is only enabled when `include`d in `$this->discoverComposerNamespaces()`. In the code, once we include `COMPOSER_PATH`, Composer will register its autoloader and prepend it in the autoload stack. If `$modules->discoverInComposer` is false, then `COMPOSER_PATH` is not included and won't be registered.

Reply:
```php
<?php

namespace App\Controllers;

use Closure;

class Home extends BaseController
{
    public function index()
    {
        foreach (spl_autoload_functions() as $autoloader) {
            switch ($autoloader) {
                case $autoloader instanceof Closure:
                    echo get_class($autoloader) . PHP_EOL;
                    break;

                case is_string($autoloader[0]):
                    echo $autoloader[0] . '::' . $autoloader[1] . PHP_EOL;
                    break;

                case is_object($autoloader[0]):
                    echo get_class($autoloader[0]) . '::' . $autoloader[1] . PHP_EOL;
                    break;
            }
        }
    }
}
```

With this PR, there is no Composer autoloader.
```
$ php public/index.php 
CodeIgniter\Autoloader\Autoloader::loadClassmap
CodeIgniter\Autoloader\Autoloader::loadClass
PHPStan\PharAutoloader::loadClass
Closure
```

Next, make `$discoverInComposer` false.

```diff
--- a/app/Config/Modules.php
+++ b/app/Config/Modules.php
@@ -29,7 +29,7 @@ class Modules extends BaseModules
      *
      * @var bool
      */
-    public $discoverInComposer = true;
+    public $discoverInComposer = false;
 
     /**
      * --------------------------------------------------------------------------
```

```
$ php public/index.php 
Composer\Autoload\ClassLoader::loadClass
CodeIgniter\Autoloader\Autoloader::loadClassmap
CodeIgniter\Autoloader\Autoloader::loadClass
PHPStan\PharAutoloader::loadClass
Closure
```
Composer autoloader is registered.

Reply:
Okay. So there's a place in the application lifecycle where Composer is included regardless of the `$discoverInComposer` flag. I'll take a look into it but most probably it shouldn't be included automatically.

Reply:
Found it!

In `system/bootstrap.php` and `system/Test/bootstrap.php`, we `require_once COMPOSER_PATH` regardless of the `$discoverInComposer` flag, which shouldn't be the case as it is handled already in the `Services::autoloader()->initialize()` call.

Reply:
```php
if (is_file(COMPOSER_PATH)) {
    require_once COMPOSER_PATH;
}
```